Update external tests

erak opened the pull request:
Closes https://github.com/ethereum/solidity/issues/6169.

Part of https://github.com/ethereum/solidity/issues/5956. This PR:
- Refactors the script the excutes external tests with the varying test setups in mind
- Re-enables the GnosisSafe contracts
- ~~Adds ColonyNetworks contracts as external project to test~~

It also adds a CircleCI job, that is triggered every night and which runs the external tests.
- [ ] Remove external tests from `build_all` in Circle config.

erak committed:
Modularize external test setup.

erak committed:
Reactivates external gnosis-safe tests.

leonardoalt commented:
I think it needs a rebase for the mac fix.
Also, is it working locally for you? Emscripten tests are failing

leonardoalt commented on test/externalTests.sh:
```diff
@@ -28,84 +28,160 @@
+  local branch="$2"
+  local dir="$3"
+
+  if [ -n "$branch" ]
```
If you pass a dir but not a branch, won't it use the dir as branch?

erak replied:
Oh, nice catch! Wanted to have the order similar for all function calls, which is why I put `$dir` to the end. I'll make the `$branch` non-optional since the projects we test do not have a identical branch names anyways.

chriseth commented on test/externalTests.sh:
```diff
@@ -28,88 +28,227 @@
+  local config_file="$1"
+  local settings="$2"
+  local evmVersion="$3"
+
```
Can you add a log statement such that we know what the tests are run against currently?

erak replied:
Yes, sure!

chriseth commented:
@leonardoalt the tests fail because of your callvalue in assembly change ;)

leonardoalt commented:
@chriseth Oh :p

chriseth commented:
That's the file that causes the failure and it looks really bad: https://github.com/dapphub/dappsys-monolithic/blob/master/note.sol

It is a logging modifier that just records `callvalue`...

chriseth commented:
Should we hot-patch the assembly code that produces the error?

erak committed:
Adds Circle job that triggers external tests nightly.

erak commented:
Updated and changed description (this PR now also proposes to execute the external tests on a nightly emscripten build instead of executing it on every branch). 

@chriseth What exactly do you mean by "doing a hotfix"? We could create a PR in the dapphub repo and use a fork until it's merged.

chriseth commented on .circleci/config.yml:
```diff
@@ -421,3 +421,17 @@ workflows:
+  test_nightly:
+    triggers:
+      - schedule:
+          cron: "0 0 * * *"
```
Is that GMT?

erak replied:
It's UTC. So this would be executed at 1:00 / 2:00 AM CET, if I'm not mistaken.

chriseth commented:
Yeah, a proper PR is of course better :)

chriseth commented:
Do you think we can get the tests to pass soon? If not, it probably makes sense to merge this split up.

erak commented:
@chriseth I've just pushed the dappsys fix and will wait for the next run. If they still fail, I'll split out Colony.

erak commented on test/externalTests.sh:
```diff
@@ -28,88 +28,238 @@
+  force_solc_truffle_modules "$SOLJSON"
+  force_solc "$CONFIG" "$DIR"
+
+  cd lib
```
This is a workaround for https://github.com/ethereum/solidity/issues/6146. I've created a PR in the 3rd party repository (https://github.com/dapphub/dappsys-monolithic/pull/2), so hopefully we can get rid of the workaround soonish.
